### PR TITLE
feat(chat): per-tool visualizers for tool calls in the conversation UI

### LINKED ISSUE #570: [Feature] Add per-tool visualizers for tool calls in the conversation UI
## Problem

The tool calls are not rendered and presented properly to the end-user. (see video)

https://github.com/user-attachments/assets/e8eaa3b1-6758-411f-8952-ff9dee2b8965

I think presenting tool calls in a comprehensible way to the end-user not only enhances the user experience, but also makes it easier to debug, perform QA, and create good snapshot tests. For these reasons, I did a bit of scoping on how we could enable that.

## Background

Tool calls (Bash, FileEditor, Grep, MCP, etc.) currently render through a single generic markdown-based path. The UI shows readable but unattractive output: bold labels, raw JSON dumps for MCP, one-line backtick commands for Bash, and no diff view for file edits.

The rendering path today:

- `src/components/conversation-events/chat/event-message-components/generic-event-message-wrapper.tsx:31` calls `getEventContent(event)` and feeds the result into `GenericEventMessage`.
- `src/components/features/chat/generic-event-message.tsx` is the collapsible card. Its `details` prop already accepts `string | React.ReactNode`, but in practice nearly everything passes markdown.
- `src/components/conversation-events/chat/event-content-helpers/get-action-content.ts` and `get-observation-content.ts` are two large `switch (kind)` blocks that build markdown strings per tool.
- The fallback (`shared.ts:5`) is `JSON.stringify(event, null, 2)` inside a fenced block.

There is precedent for React-component rendering: `TaskTrackerObservation` already returns `<TaskTrackingObservationContent>` instead of markdown (`get-event-content.tsx:296`). This issue generalizes that pattern.

## Worst offenders today

| Tool | Current rendering |
|---|---|
| `MCPToolAction` (`get-action-content.ts:103`) | Raw `JSON.stringify(action.data, null, 2)` inside a code block |
| `ExecuteBashAction` (`get-action-content.ts:89`) | `Command:\n\`${cmd}\`` — single line, no syntax highlight, no copy |
| `FileEditorAction` (`get-action-content.ts:68`) | Only `create` shows file content; `edit` / `str_replace` / `view` show nothing useful |
| `GrepAction` / `GlobAction` (`get-action-content.ts:50`) | Bold labels for params; matches rendered as plain markdown bullet list |
| Default fallback (`shared.ts:5`) | Full event serialized as JSON |

## Goals

1. Replace the markdown-string output of high-traffic tools with dedicated React visualizers.
2. Make adding a new visualizer cost one file and one barrel-import line.
3. Preserve the existing card chrome, collapse behavior, success indicator, and ACP rendering path.
4. Keep the markdown path as a fallback so unmigrated tools keep working unchanged.

## Non-goals

- No changes to titles, i18n keys, the success indicator, confirmation buttons, or ACP tool-call rendering.
- No new dependencies — hand-rolled with Tailwind, matching current conventions.
- No combined action+observation cards in this PR. Two-card layout (action card, then observation card) stays as-is.
- No migration of low-impact tools (Think, Finish, browser ops, MCP, InvokeSkill, TaskTracker). Those follow in later PRs.

### Type-safe registry via a `defineVisualizer` helper

A single helper that narrows `Action` / `Observation` by their literal `kind` so visualizer bodies receive correctly typed props without casts:

```ts
// src/components/features/chat/tool-visualizers/define.ts
type ActionKind      = Action["kind"];
type ObservationKind = Observation["kind"];
type ActionByKind<K extends ActionKind>           = Extract<Action, { kind: K }>;
type ObservationByKind<K extends ObservationKind> = Extract<Observation, { kind: K }>;

export interface VisualizerProps<A extends Action, O extends Observation> {
  action: ActionEvent<A>;
  observation?: ObservationEvent<O>;
}

export interface ToolVisualizer<AK extends ActionKind, OK extends ObservationKind> {
  actionKind: AK;
  observationKind?: OK;
  titleKey: string;
  icon?: React.ComponentType<{ className?: string }>;
  Body: React.FC<VisualizerProps<ActionByKind<AK>, ObservationByKind<OK>>>;
}

export const defineVisualizer = <AK extends ActionKind, OK extends ObservationKind>(
  v: ToolVisualizer<AK, OK>,
) => v;
```

### Explicit barrel registration

```ts
// src/components/features/chat/tool-visualizers/index.ts
import { bashVisualizer }       from "./bash/bash";
import { fileEditorVisualizer } from "./file-editor/file-editor";
import { searchVisualizer }     from "./search/search";

const ALL = [bashVisualizer, fileEditorVisualizer, searchVisualizer];

export const actionVisualizers      = new Map(ALL.map(v => [v.actionKind, v]));
export const observationVisualizers = new Map(
  ALL.filter(v => v.observationKind).map(v => [v.observationKind!, v]),
);
```

Auto-discovery via `import.meta.glob` was considered and rejected: it hurts tree-shaking, breaks test mocking, and hides what is wired.

### Dispatch seam

`get-event-content.tsx` checks the registry before falling through to the existing markdown helpers:

```tsx
function resolveVisualizer(event, correspondingAction) {
  if (isActionEvent(event)) {
    const v = actionVisualizers.get(event.action.kind);
    if (v) return { title: t(v.titleKey), body: <v.Body action={event} /> };
  }
  if (isObservationEvent(event) && correspondingAction) {
    const v = observationVisualizers.get(event.observation.kind);
    if (v) return { title: t(v.titleKey), body: <v.Body action={correspondingAction} observation={event} /> };
  }
  return null;
}
```

If `resolveVisualizer` returns null, the current markdown pipeline runs unchanged.

### Shared primitives

Dumb, presentation-only components in `tool-visualizers/primitives/`. No knowledge of event types. Each is a few dozen lines of Tailwind:

- `CodeBlock` — monospace, language hint, copy button, expand when long.
- `FilePathChip` — monospace path with copy icon and optional line range.
- `DiffView` — unified diff for `str_replace`-style edits.
- `OutputPane` — stdout / stderr split, exit-code badge, collapsible.
- `KeyValueGrid` — for simple key/value displays.

### Example visualizer (full file)

```tsx
// src/components/features/chat/tool-visualizers/bash/bash.tsx
import TerminalIcon from "#/icons/terminal.svg?react";
import { defineVisualizer } from "../define";
import { CodeBlock } from "../primitives/code-block";
import { OutputPane } from "../primitives/output-pane";

export const bashVisualizer = defineVisualizer({
  actionKind: "ExecuteBashAction",
  observationKind: "ExecuteBashObservation",
  titleKey: "ACTION_MESSAGE$RUN",
  icon: TerminalIcon,
  Body: ({ action, observation }) => (
    <div className="flex flex-col gap-2">
      <CodeBlock language="bash" copy>{action.action.command}</CodeBlock>
      {observation && (
        <OutputPane
          stdout={observation.observation.output}
          exitCode={observation.observation.exit_code}
        />
      )}
    </div>
  ),
});
```

## Folder layout

```
src/components/features/chat/tool-visualizers/
  define.ts
  index.ts
  dispatcher.ts
  test-utils.tsx
  primitives/
    code-block.tsx
    file-path-chip.tsx
    diff-view.tsx
    output-pane.tsx
    key-value-grid.tsx
  bash/
    bash.tsx
    bash.test.tsx
  file-editor/
    file-editor.tsx
    file-editor.test.tsx
  search/
    search.tsx
    search.test.tsx
```

## Scope of this issue (Tier 1)

Migrate the three highest-traffic, ugliest-today tools, each as an action + observation pair:

| Tool | Action | Observation |
|---|---|---|
| Bash | `ExecuteBashAction` | `ExecuteBashObservation` |
| File editor | `FileEditorAction`, `StrReplaceEditorAction` | `FileEditorObservation`, `StrReplaceEditorObservation` |
| Search | `GrepAction`, `GlobAction` | `GrepObservation`, `GlobObservation` |

### Visual targets

| Tool | Today | After |
|---|---|---|
| Bash | One-line backtick command, raw stdout dump | `CodeBlock` (bash) for command, `OutputPane` with stdout/stderr split and exit-code badge |
| FileEditor `create` | Path + truncated text | `FilePathChip` + `CodeBlock` with detected language |
| FileEditor `str_replace` | Nothing | `DiffView` of old vs new |
| FileEditor `view` | Nothing | `FilePathChip` + line-range badge |
| Grep / Glob | Bold labels + bullet list | Pattern and path chips, match count, file:line list |

## How to add a new visualizer (post-merge)

Documented in `tool-visualizers/index.ts`:

1. Create `tool-visualizers/<name>/<name>.tsx` exporting `defineVisualizer({ ... })`.
2. Add one import and one entry in the `ALL` array in `index.ts`.
3. Add `<name>.test.tsx` using `renderVisualizer` with fixtures.

TypeScript enforces the rest: `actionKind` autocompletes from the `Action` union, the `Body` component receives a narrowed event type, and typos in kinds are compile errors.

## Testing

- Co-locate `*.test.tsx` next to each visualizer.
- Reuse fixtures already in `__tests__/components/conversation-events/chat/`.
- One snapshot per (action, observation) pair, plus error states for Bash (non-zero exit) and FileEditor (failed edit).
- Existing tests in `__tests__/components/conversation-events/chat/event-message-*.test.tsx` and `__tests__/utils/handle-event-for-ui.test.ts` must continue to pass for unmigrated tools (fallback path).

## Risks and edge cases

- The markdown strings are also consumed by ACP tool-call rendering (`get-event-content.tsx:305`), which builds details "the same way" as bash output. The dispatcher must only intercept the action/observation `details` body, not the title pipeline or the ACP path.
- `correspondingAction` may be undefined for an observation event (e.g., resumed conversation, missing action). The dispatcher falls back to markdown in that case.
- Streaming: action arrives before observation. Visualizers render the action card immediately; the observation card appears later. Two-card layout means no coordination needed.

## Acceptance criteria

- [ ] `tool-visualizers/` directory exists with `define.ts`, `index.ts`, `dispatcher.ts`, `test-utils.tsx`, and the five primitives.
- [ ] Bash, FileEditor (all four commands), and Grep/Glob render via the new visualizers, both action and observation cards.
- [ ] All other tools render unchanged (markdown fallback).
- [ ] Snapshot tests cover the migrated visualizers, including error states.
- [ ] Existing tests in `__tests__/` pass without modification.
- [ ] README or in-file comment in `tool-visualizers/index.ts` documents the three-step recipe for adding a new visualizer.

## Follow-ups (not in this issue)

- Tier 2: MCP, InvokeSkill, main Browser actions.
- Tier 3: Think, Finish, remaining browser ops, TaskTracker polish.
- Optional: combined action+observation cards behind a per-visualizer opt-in.

### LINKED ISSUE #864: [Bug]: Subagents not rendering nicely
### Operating System

macOS

### Installation Method

Clone and `npm run dev`

### Agent Canvas Version

main

### Bug Description

Image

Says TASK instead of something more descriptive

### Steps to Reproduce

* enable subagents in settings
* "send off two subagents to flip a coin and report back"
* watch chat

### Actual Behavior

says TASK

### Expected Behavior

Says something more descriptive

### Relevant Logs

```shell

```

### Screenshots

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by VascoSch92**
- [x] A human has tested these changes.
  
  ---

  ## Why
  
  Tool calls were all rendered through one generic markdown pipeline, so every
  tool looked alike and tools the frontend didn't model (e.g. `task` and
  `canvas_ui`) fell back to dumping the raw event JSON. This adds purpose-built,
  per-tool visualizers that render a tool call's action/observation card as React.
  The system is incremental: unregistered tools keep the markdown fallback, so
  visualizers can be added one tool at a time.

  ## Summary
  
  - **Framework**: a type-safe visualizer registry (`defineVisualizer` + dispatcher)
    with shared primitives (code block, diff view, key/value grid, output pane),
    plus visualizers for **bash/terminal**, **file editor**, and **search**.
  - **Task (subagent) visualizer**: shows the subagent, task id, the **query the
    parent agent sent**, and the subagent's markdown result. While the task is
    in-flight (no observation yet) it shows just the query.
  - **Invoke-skill** results now render in the existing "Skill Ready" expandable
    list under an "Invoked Skill Knowledge" header (reuses `SkillReadyContentList`).
  - **`canvas_ui`** observations now show their acknowledgement text instead of a
    raw JSON dump.
  - **UX**: hover copy buttons on the command, output, and subagent result; removed
    the decorative icons from the bash/terminal card. Success/error dots now honor
    `is_error`/`status` for the invoke-skill, task, and canvas observations.

  ## Issue Number

  Closes #570

  ## How to Test

```
  npm install
  npm test          # 3066 unit tests pass
  npm run typecheck  # clean
```

  Manual: `npm run dev` (or `npm run dev:mock`), open a conversation and trigger
  bash/grep/file edits and a subagent `task`, then confirm each tool renders a
  dedicated card, the copy buttons work on command/output/result, and an in-flight
  `task` shows the query before its result arrives.

  ## Video/Screenshots

 
Screenshot 2026-06-08 at 13 43 55
Screenshot 2026-06-08 at 13 45 01
Screenshot 2026-06-08 at 13 46 45
Screenshot 2026-06-08 at 13 47 20
Screenshot 2026-06-08 at 13 47 57

  ## Type
  
  - [ ] Bug fix
  - [x] Feature
  - [ ] Refactor
  - [ ] Breaking change
  - [ ] Docs / chore

  ## Notes

  - Adds i18n keys across all 15 locales (translation-completeness gate passes).
  - Models a few events client-side that the frontend hadn't before: `TaskAction`/
    `TaskObservation` and `CanvasUIObservation`. `TaskAction`'s schema (`prompt`,
    `subagent_type`, …) was verified against the SDK source, not guessed.
  - Bundles a `main` merge; the only conflict was `package-lock.json`, resolved to
    the `@openhands/extensions` 0.3.0 npm release (matching the merged `package.json`).
  - Follow-up: no full-render test of `GenericEventMessageWrapper` for the
    invoke-skill toggle (its parts are unit-tested individually).

---
**🐳 Docker images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-canvas/pkgs/container/agent-canvas

| Component | Value |
|---|---|
| **Image** | `ghcr.io/openhands/agent-canvas` |
| **Architectures** | amd64, arm64 |
| **Agent Server** | `ghcr.io/openhands/agent-server:1.26.0-python` |
| **Automation** | `openhands-automation==1.0.0a6` |
| **Commit** | `d50d64be06837458bc296b879f648d74d11410a9` |

**Pull (multi-arch manifest)**
```bash
# Multi-arch manifest — Docker automatically pulls the correct architecture
docker pull ghcr.io/openhands/agent-canvas:sha-d50d64b
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  ghcr.io/openhands/agent-canvas:sha-d50d64b
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-canvas:sha-d50d64b-amd64
ghcr.io/openhands/agent-canvas:570-feature-add-per-tool-visualizers-for-tool-calls-in-the-conversation-ui-amd64
ghcr.io/openhands/agent-canvas:pr-1246-amd64
ghcr.io/openhands/agent-canvas:sha-d50d64b-arm64
ghcr.io/openhands/agent-canvas:570-feature-add-per-tool-visualizers-for-tool-calls-in-the-conversation-ui-arm64
ghcr.io/openhands/agent-canvas:pr-1246-arm64
ghcr.io/openhands/agent-canvas:sha-d50d64b
ghcr.io/openhands/agent-canvas:570-feature-add-per-tool-visualizers-for-tool-calls-in-the-conversation-ui
ghcr.io/openhands/agent-canvas:pr-1246
```

**About Multi-Architecture Support**
- Each tag (e.g., `sha-d50d64b`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `sha-d50d64b-amd64`) are also available if needed

### CODE REVIEW

File: src/i18n/translation.json [hunk 1]
```diff
@@ -1,4 +1,225 @@
...
+    "fr": "Results truncated",
+    "it": "Results truncated",
+    "ja": "Results truncated",
+    "ko-KR": "Results truncated",
+    "no": "Results truncated",
+    "pt": "Results truncated",
+    "tr": "Results truncated",
+    "uk": "Results truncated",
+    "zh-CN": "Results truncated",
+    "zh-TW": "Results truncated"
```
Comment (hieptl) [inline src/i18n/translation.json]: **suggestion (blocking):** We need to support multiple languages. Thank you! 🙏

**Review by hieptl — CHANGES_REQUESTED** [review]

**Comment by enyst** [comment]
@openhands Do a /codereview on this PR. Understand the goal and implications. Post your feedback as a review in the PR. Don’t be too mean on nits.

**Comment by enyst** [comment]
@VascoSch92 Please feel free to consider or have your agent consider my agent’s feedback or ignore it as you see fit. I’m just trying to get GPT-5.5 to surface tricky stuff if there is any, but sometimes it exagerates!

**Review by enyst — COMMENTED** [review]
🟡 **Taste Rating: Acceptable** — the visualizer registry is a reasonable incremental seam and the tests/CI coverage are solid, but I found a couple of correctness gaps in the newly added rendering/history behavior.

_This review was generated by an AI agent (OpenHands) on behalf of the requester._

**[CRITICAL ISSUES]**
- None.

**[IMPROVEMENT OPPORTUNITIES]**
- [`src/components/features/chat/tool-visualizers/file-editor/file-editor.tsx:41`](https://github.com/OpenHands/agent-canvas/pull/1246/files) **Correctness:** the file-editor diff path uses truthiness (`obs.old_content && obs.new_content`), so valid edits where one side is an empty string skip the diff entirely. That includes clearing a file (`new_content: ""`) or inserting into an empty file (`old_content: ""`). The fallback then either shows only the output text or nothing useful, which undercuts the purpose of this visualizer for real edit cases. Use nullish checks (`!= null`) instead, and add tests for empty old/new content.
- [`src/components/features/chat/tool-visualizers/file-editor/file-editor.tsx:68`](https://github.com/OpenHands/agent-canvas/pull/1246/files) **In-flight insert rendering:** pending `insert` actions require `old_str != null`, but normal insert calls do not have an `old_str`; they have `new_str` plus `insert_line`. As a result, in-flight insert actions render only the path and hide the inserted text. Render inserts from `new_str` without requiring `old_str` (for example as an insertion diff or labeled code block) and cover it in the visualizer tests.
- [`src/contexts/conversation-websocket-context.tsx:580`](https://github.com/OpenHands/agent-canvas/pull/1246/files) / [`src/stores/model-store.ts:105`](https://github.com/OpenHands/agent-canvas/pull/1246/files) **State consistency:** successful `SwitchLLMObservation`s now get hidden and replaced with inline model messages, but the agent-triggered path only records the visual message and updates the model cache. It does not persist `active_profile` the way the user-driven switch path does, so after reload the header can fall back to model-based matching and show the wrong profile when multiple profiles share a model. Also, `seedSwitches` only runs for the initial REST preload; if an older page loaded by `useLoadOlderEvents` contains a successful switch, the underlying observation remains hidden and no model message is seeded for it. Please persist the profile on successful agent-triggered switches and either reseed/update switch messages after older pages are merged or derive the anchors from current `uiEvents` at render time.

**[TESTING GAPS]**
- Add focused tests for the cases above: file-editor empty-string diffs, pending insert actions, and a model-switch observation that arrives through older-history pagination. The existing tests cover the happy path but would not catch these regressions.

**[RISK ASSESSMENT]**
- [Overall PR] ⚠️ Risk Assessment: 🟡 MEDIUM. This is frontend-only, behind existing event data, and CI/e2e are green. The risk is still medium because it changes the chat's core event rendering path and hides some events in favor of reconstructed UI state; losing history messages or hiding edit details is user-visible and can make debugging conversations harder.

**VERDICT:**
❌ **Needs follow-up:** The architecture is sound, but I would address the empty edit/insert rendering and model-switch history consistency before merging.

**KEY INSIGHT:**
The visualizer seam is good, but once a raw event is hidden in favor of derived UI, every history-loading path has to rebuild that derived UI or the transcript silently loses information.

---

> **Improve this review?** If any feedback above seems incorrect or irrelevant to this repository, you can teach the reviewer to do better:
>
> 1. Add a `.agents/skills/custom-codereview-guide.md` file to your branch (or edit it if one already exists) with the `/codereview` trigger and the context the reviewer is missing (e.g., "Security concerns about X do not apply here because Y"). See the [customization docs](https://docs.openhands.dev/openhands/usage/use-cases/code-review#customization) for the required frontmatter format.
> 2. Re-request a review - the reviewer reads guidelines from the PR branch, so your changes take effect immediately.
> 3. When your PR is merged, the guideline file goes through normal code review by repository maintainers.
>
> **Resolve with AI?** Install the [iterate skill](https://github.com/OpenHands/extensions/tree/main/skills/iterate) in your agent and run `/iterate` to automatically drive this PR through CI, review, and QA until it's merge-ready.
>
> Was this review helpful? React with 👍 or 👎 to give feedback.

**Comment by enyst** [comment]
@openhands Understand the PR and help us visualize and understand how it works when an event is rendered this way: clone yourself enyst.github.io and add a new page under arch to show the flow/design and make sure you explain and show in diagrams how it works to render and event; make sure the page renders correctly. (push it out for us to see and give us a link in a github comment)

Add another section to that page that answers the question: how does a user define a different way to render an event or more? If they cannot do it without editing their fork, then investigate and propose a simple way for a followup arch that will allow people to add addons that define new visualizers and plug them in without editing code.

Find the open issue posted on agent-canvas by devin, where he proposes addons; understand all comments in that proposal and post a comment there where you link these and propose how to add custom visualizers.

**Comment by enyst** [comment]
Done.

- Added and published the architecture page here: https://enyst.github.io/arch/agent-canvas-tool-visualizers.html
- Commit in `enyst.github.io`: https://github.com/enyst/enyst.github.io/commit/f105ad9
- Verified the published page returns 200 and the homepage links to it.
- Found Devin's addon proposal issue (#481) and posted the custom visualizer/addon proposal here: https://github.com/OpenHands/OpenHands/issues/15949#issuecomment-4658100889

The page explains the current render flow with diagrams, including:

- action/observation events entering through REST history / WebSocket into the conversation renderer;
- `GenericEventMessageWrapper` -> `getEventContent()` -> `resolveVisualizerBody()`;
- registry lookup by action/observation `kind`;
- React visualizer body vs markdown fallback;
- how the observation card finds the source action via `action_id`;
- what is supported today for bash/terminal, file editor, search, and task/subagent events;
- the current limitation that users must edit/fork Agent Canvas to add a renderer;
- a follow-up architecture for addon-provided visualizers via `registerToolVisualizer()` / `conversation.event.visualizers`.

_This comment was created by an AI agent (OpenHands) on behalf of the user._

**Review by enyst — APPROVED** [review]
LGTM, thank you!

**Review by hieptl — APPROVED** [review]
Thank you! 🙏
